### PR TITLE
Tensorboard logging/visualization

### PR DESCRIPTION
**Pull request by filipeabperes**
This is something I've been using, but I'm not sure if it should be merged to master. Since it could be useful to some, I'm placing it here. It's a modification to the train script that sets up a flag for allowing tensorboard logging, which can then be visualized (even remotely) with a tensorboard server . The train script only sets up the accuracy, but the agent can set up other values for logging by placing them in the opt['tensorboard'] dict, eg below.

image

### CODE REVIEW

**Comment by belbs** [comment]
yeah, I thought it seemed too hacky (?) to merge into the current script as well, which is pretty clean. not sure where to put it though, but thought someone might benefit from it just by putting it over here as a separate branch

**Comment by alexholdenmiller** [comment]
we could add a logger to the utils which handles printing / logging to tensorboard / w.e depending on params?

**Comment by alexholdenmiller** [comment]
then just simple log command in the train_model loop and the logger will do all the fancy stuff

**Comment by jaseweston** [comment]
that would be cool

**Comment by belbs** [comment]
that could work, but I'm using a third party library to log in tensorboard format. the reference to it could just be moved into the logger, I suppose (?)

**Comment by alexholdenmiller** [comment]
right!

**Comment by belbs** [comment]
I moved the logger into a class in utils. Once it is created I keep a reference to it in opt, so that models can log data as well, eg below with the loss. (But the logger is not copied on deepcopy of opt, in order not to waste space/time.) 

```
if 'logger' in self.opt:
            self.opt['logger'].log_train({'loss': loss.data[0]})
```

File: examples/train_model.py [hunk 1]
```diff
@@ -93,13 +94,20 @@ def main():
...
     # Possibly build a dictionary (not all models do this).
     if opt['dict_build_first'] and 'dict_file' in opt:
         if opt['dict_file'] is None and opt.get('model_file'):
             opt['dict_file'] = opt['model_file'] + '.dict'
         print("[ building dictionary first... ]")
         build_dict.build_dict(opt)
+
+    # Set up logger if directory provided
+    if opt['tensorboard_dir']:
+        opt['logger'] = TensorboardLogger(opt['tensorboard_dir'])
```
Comment (jaseweston) [inline examples/train_model.py]: we are trying to keep opt as only strings, float, bools etc,  not whole classes.

**Comment by belbs** [comment]
Though it doesn't seem very clean, the only other way I can think of to pass values between the training script and the actual model is via a dictionary in opt (the way I put it now).

**Comment by alexholdenmiller** [comment]
so does the agent need to fill that log dict then--that info isn't available at this level of the loop? (we don't have your agent's side of it in this diff--what is it putting in that dict exactly?)

one other option other than using "opt" would be to pass it in the "shared" table to the agent but maybe that's too hacky too

**Comment by belbs** [comment]
ah, yes, forgot to provide an example for that. eg, the agent would do:

```
if 'train_log' in self.opt:
            self.opt['train_log']['loss'] = loss.data[0]
```

And yes the shared is another possibility. In general I think it is good to think about this because it might be useful to have a defined way for the train loop to get information from the agent. Even if the type of logger is swapped later, this would still be useful.

**Comment by jaseweston** [comment]
opt is meant to be for options..

**Comment by jaseweston** [comment]
closing this task as we have a new, fresh one coming!
